src/app/application/components/data-source/DataSource.vue

```diff
@@ -0,0 +1,78 @@
+<script setup lang="ts">
+
+import { watch, ref, onBeforeUnmount } from 'vue'
+
+import { useDataSource } from '@/utilities'
+
+const data = useDataSource()
+
+const props = defineProps({
+  src: {
+    type: String,
+    required: true,
+  },
+})
+
+const message = ref<unknown | undefined>(undefined)
+const error = ref<Error | undefined>(undefined)
+
+const emit = defineEmits<{
+  (e: 'change', value: any): void
+  (e: 'error', error: Error): void
+}>()
+
+let controller: AbortController
+let _src: string
+const sym = Symbol('')
+const open = async (src: string) => {
+  close(controller, _src)
+  _src = src
+  controller = new AbortController()
+  // this should emit proper events
+  const source = data.source(src, sym)
+  if (typeof source !== 'undefined') {
+    source.addEventListener(
+      'message',
+      (e: Event) => {
+        message.value = (e as MessageEvent).data
+        emit('change', message.value)
+      },
+      { signal: controller.signal },
+    )
+    source.addEventListener(
+      'error',
+      (e: Event) => {
+        error.value = (e as ErrorEvent).error as Error
+        emit('error', error.value)
+      },
+      { signal: controller.signal },
+    )
+  }
+}
+const close = (controller: AbortController | undefined, src: string) => {
+  if (typeof controller !== 'undefined') {
+    controller.abort()
+  }
+  if (typeof src !== 'undefined') {
+    data.close(src, sym)
+  }
+  controller = undefined
+}
+watch(() => props.src, (src) => open(src), { immediate: true })
+onBeforeUnmount(() => {
+  close(controller, _src)
+})
+const refresh = () => {
+  close(controller, _src)
+  open(props.src)
+}
+</script>
+
+<template>
+  <slot
+    :data="message as any"
+    :error="error"
+    :refresh="refresh"
+  />
+  <data />
```

Comment:
This `data` HTML element will allow us to do lazy loading.

Reply:
Oh weird I thought I'd added an inline comment here, it's so we can do lazy loading i.e. so we know when the DataSource is in the viewport. Can be any HTML element really, but I've always considered populating the `data` element with the actual data, then it would work with native `select`/`data` interactions (not saying I'm gonna do that at all.

Reply:
So the intention is to not actually populate this element whatsoever but rather just make it "browser visible" so that, potentially, mechanisms like IntersectionObserver can do visibility checks?

If so, then this should definitely just be a `<div />` I think.

Reply:
`div` s are more likely to get styled by accident. `data` not so much. I actually don't think it matters that much apart from that, plus the fact that its in one place and takes all of 2 seconds to change with not a lot of impact on the codebase.

All in all, unless you can think of any other downsides to using `data` I'd rather keep it as that if only to be less likely to be accidentally styled.

Reply:
>So the intention is to not actually populate this element whatsoever but rather just make it "browser visible" so that, potentially, mechanisms like IntersectionObserver can do visibility checks?

At the moment, yes we won't be populating it right now. But if we want to we can, as I said it might be useful for native `select/data` and other interactions. but yes, IntersectionObserver for the moment (which we don't have here but probably will add later when we first need it)

Reply:
> `div` s are more likely to get styled by accident. `data` not so much.

How? We control that and we don’t have a single instance of selecting a descendant `div` in our code base.

It should just be a `div` because it’s completely non-semantic, never will carry any actual data and we also won’t set the `data` element’s `value` attribute. There is no reason to use `data` here.

Reply:
Couple of things here. `DataSource` is a representation of data and div is a representation of nothing/everything, so something that says 'data' makes more sense semantically.

Secondly I think I actually meant to use `datalist` here 😆 🤷 . Honestly don't really care right now, its in one place so its easy to change later if we want than spend time discussing semantics here.

Just want to point out there is no technical reason that I'm aware of as to why using `data` is detrimental (but I've given a reason why a div may _become_ one), tbh feels like opinion/bikeshedding more than anything 🤷 

All in all I changed this to a `div`